src/ring_theory/adjoin/basic.lean

```diff
@@ -102,6 +113,65 @@ lemma adjoint_prod_le (s : set A) (t : set B) :
   adjoin R (set.prod s t) ≤ (adjoin R s).prod (adjoin R t) :=
 adjoin_le $ set.prod_mono subset_adjoin subset_adjoin
 
+lemma adjoin_le_prod (s) (t) :
+  adjoin R (linear_map.inl R A B '' (s ∪ {1}) ∪
+  linear_map.inr R A B '' (t ∪ {1})) ≤ (adjoin R s).prod (adjoin R t) :=
+begin
+  refine adjoin_le (set.union_subset _ _),
+  { rw set.image_union,
+    refine set.union_subset (λ x hx, set_like.mem_coe.2 $ set.mem_prod.2 _)
+      (λ x hx, set_like.mem_coe.2 $ set.mem_prod.2 _),
```

Comment:
Would a `subalgebra.mem_prod` lemma make this slightly shorter?

Reply:
We have `subalgebra.mem_prod`, but I don't think it can help. Given `x` in `adjoin R (linear_map.inl R A B '' (s ∪ {1}) ∪
  linear_map.inr R A B '' (t ∪ {1}))` we don't know that much (before the lemma we are proving) about `x.1` and `x.2`.

The proof seems long, but it is so just because we are talking about the `adjoin` of the union of four sets, so we have in a way or in another four cases to treat.

Reply:
If you see a nice `rcases` use (or something else) after

```lean
lemma adjoin_le_prod (s) (t) :
  adjoin R (linear_map.inl R A B '' (s ∪ {1}) ∪
  linear_map.inr R A B '' (t ∪ {1})) ≤ (adjoin R s).prod (adjoin R t) :=
begin
  rw [set.image_union, set.image_union],
  refine adjoin_le _,

end
```
to have the four cases as separate goals, the proof will become probably a little nice.

Reply:
Ah, it's `rcases hx with (a | b) | (c | d)`

Reply:
Now it's better.